File: benchmark/quicksort/sort.lua
local function swap(arr, i, j)
    arr[i], arr[j] = arr[j], arr[i]
end

local function partition(arr, low, high)
    local pivot = arr[high]
    local i = low - 1
    for j = low, high - 1 do
        if arr[j] <= pivot then
            i = i + 1
            swap(arr, i, j)
        end
    end
    swap(arr, i + 1, high)
    return i + 1
end

local function quicksort(arr, low, high)
    if low < high then
        local pi = partition(arr, low, high)
        quicksort(arr, low, pi - 1)
        quicksort(arr, pi + 1, high)
    end
end

local function isSorted(arr, n)
    for i = 2, n do
        if arr[i - 1] > arr[i] then
            return false
        end
    end
    return true
end

local function main()
    local size = 1000
    local iterations = 10
    local checksum = 0

    for iter = 0, iterations - 1 do
        -- Create array with pseudo-random values
        local arr = {}
        for i = 1, size do
            -- Simple LCG: (a * x + c) mod m (using i-1 for 0-based logic)
            arr[i] = (((i - 1) * 1103515245 + 12345 + iter) % 1000000) - 500000
        end

        quicksort(arr, 1, size)

        if isSorted(arr, size) then
            checksum = checksum + arr[1] + arr[size]
        end
    end

    print(checksum)
end

main()

File: benchmark/quicksort/sort.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

void swap(int *arr, int i, int j) {
    int tmp = arr[i];
    arr[i] = arr[j];
    arr[j] = tmp;
}

int partition(int *arr, int low, int high) {
    int pivot = arr[high];
    int i = low - 1;
    for (int j = low; j < high; j++) {
        if (arr[j] <= pivot) {
            i++;
            swap(arr, i, j);
        }
    }
    swap(arr, i + 1, high);
    return i + 1;
}

void quicksort(int *arr, int low, int high) {
    if (low < high) {
        int pi = partition(arr, low, high);
        quicksort(arr, low, pi - 1);
        quicksort(arr, pi + 1, high);
    }
}

bool isSorted(int *arr, int n) {
    for (int i = 1; i < n; i++) {
        if (arr[i - 1] > arr[i]) {
            return false;
        }
    }
    return true;
}

int main() {
    int size = 3000;
    int iterations = 20;
    long checksum = 0;

    for (int iter = 0; iter < iterations; iter++) {
        // Create array with pseudo-random values
        int *arr = (int *)malloc(size * sizeof(int));
        for (int i = 0; i < size; i++) {
            // Simple LCG: (a * x + c) mod m
            arr[i] = (((long)i * 1103515245 + 12345 + iter) % 1000000) - 500000;
        }

        quicksort(arr, 0, size - 1);

        if (isSorted(arr, size)) {
            checksum += arr[0] + arr[size - 1];
        }

        free(arr);
    }

    printf("%ld\n", checksum);
    return 0;
}
